README.md
## `gluufederation/key-rotation:4.1.1_06` is the last supported version for this image. This image has been depreciated post `4.1.1_06` in favor for [gluufederation/docker-certmanager](https://github.com/GluuFederation/docker-certmanager).

## Overview

KeyRotation is a special container to rotate OpenID keys for oxAuth.

## Versions

- Stable: `gluufederation/key-rotation:4.1.1_06`
- Unstable: `gluufederation/key-rotation:4.1.1_dev`

Refer to [Changelog](./CHANGES.md) for details on new features, bug fixes, or older releases.

## Environment Variables

The following environment variables are supported by the container:

- `GLUU_CONFIG_ADAPTER`: The config backend adapter, can be `consul` (default) or `kubernetes`.
- `GLUU_CONFIG_CONSUL_HOST`: hostname or IP of Consul (default to `localhost`).
- `GLUU_CONFIG_CONSUL_PORT`: port of Consul (default to `8500`).
- `GLUU_CONFIG_CONSUL_CONSISTENCY`: Consul consistency mode (choose one of `default`, `consistent`, or `stale`). Default to `stale` mode.
- `GLUU_CONFIG_CONSUL_SCHEME`: supported Consul scheme (`http` or `https`).
- `GLUU_CONFIG_CONSUL_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_CONFIG_CONSUL_CACERT_FILE`: path to Consul CA cert file (default to `/etc/certs/consul_ca.crt`). This file will be used if it exists and `GLUU_CONFIG_CONSUL_VERIFY` set to `true`.
- `GLUU_CONFIG_CONSUL_CERT_FILE`: path to Consul cert file (default to `/etc/certs/consul_client.crt`).
- `GLUU_CONFIG_CONSUL_KEY_FILE`: path to Consul key file (default to `/etc/certs/consul_client.key`).
- `GLUU_CONFIG_CONSUL_TOKEN_FILE`: path to file contains ACL token (default to `/etc/certs/consul_token`).
- `GLUU_CONFIG_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_CONFIG_KUBERNETES_CONFIGMAP`: Kubernetes configmaps name (default to `gluu`).
- `GLUU_CONFIG_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_SECRET_ADAPTER`: The secrets adapter, can be `vault` or `kubernetes`.
- `GLUU_SECRET_VAULT_SCHEME`: supported Vault scheme (`http` or `https`).
- `GLUU_SECRET_VAULT_HOST`: hostname or IP of Vault (default to `localhost`).
- `GLUU_SECRET_VAULT_PORT`: port of Vault (default to `8200`).
- `GLUU_SECRET_VAULT_VERIFY`: whether to verify cert or not (default to `false`).
- `GLUU_SECRET_VAULT_ROLE_ID_FILE`: path to file contains Vault AppRole role ID (default to `/etc/certs/vault_role_id`).
- `GLUU_SECRET_VAULT_SECRET_ID_FILE`: path to file contains Vault AppRole secret ID (default to `/etc/certs/vault_secret_id`).
- `GLUU_SECRET_VAULT_CERT_FILE`: path to Vault cert file (default to `/etc/certs/vault_client.crt`).
- `GLUU_SECRET_VAULT_KEY_FILE`: path to Vault key file (default to `/etc/certs/vault_client.key`).
- `GLUU_SECRET_VAULT_CACERT_FILE`: path to Vault CA cert file (default to `/etc/certs/vault_ca.crt`). This file will be used if it exists and `GLUU_SECRET_VAULT_VERIFY` set to `true`.
- `GLUU_SECRET_KUBERNETES_NAMESPACE`: Kubernetes namespace (default to `default`).
- `GLUU_SECRET_KUBERNETES_CONFIGMAP`: Kubernetes secrets name (default to `gluu`).
- `GLUU_SECRET_KUBERNETES_USE_KUBE_CONFIG`: Load credentials from `$HOME/.kube/config`, only useful for non-container environment (default to `false`).
- `GLUU_WAIT_MAX_TIME`: How long the startup "health checks" should run (default to `300` seconds).
- `GLUU_WAIT_SLEEP_DURATION`: Delay between startup "health checks" (default to `10` seconds).
- `GLUU_KEY_ROTATION_INTERVAL`: how long the key should be expired (default to 48 hours).
- `GLUU_KEY_ROTATION_CHECK`: delay between rotation check (default to 3600 seconds).
- `GLUU_KEY_ROTATION_FORCE`: force key rotation regardless of `GLUU_KEY_ROTATION_INTERVAL` value (default to `false`).
- `GLUU_PERSISTENCE_TYPE`: Persistence backend being used (one of `ldap`, `couchbase`, or `hybrid`; default to `ldap`).
- `GLUU_PERSISTENCE_LDAP_MAPPING`: Specify data that should be saved in LDAP (one of `default`, `user`, `cache`, `site`, or `token`; default to `default`). Note this environment only takes effect when `GLUU_PERSISTENCE_TYPE` is set to `hybrid`.
- `GLUU_LDAP_URL`: Address and port of LDAP server (default to `localhost:1636`); required if `GLUU_PERSISTENCE_TYPE` is set to `ldap` or `hybrid`.
- `GLUU_COUCHBASE_URL`: Address of Couchbase server (default to `localhost`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_USER`: Username of Couchbase server (default to `admin`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_CERT_FILE`: Couchbase root certificate location (default to `/etc/certs/couchbase.crt`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_COUCHBASE_PASSWORD_FILE`: Path to file contains Couchbase password (default to `/etc/gluu/conf/couchbase_password`); required if `GLUU_PERSISTENCE_TYPE` is set to `couchbase` or `hybrid`.
- `GLUU_CONTAINER_METADATA`: The name of scheduler to pull container metadata (one of `docker` or `kubernetes`; default to `docker`).


CHANGES.md
# Changelog

Here you can see an overview of changes between each release.

## Version 3.1.6_04

* Fixed mismatched keystore password between secret and persistence (after rotating keys).

## Version 3.1.6_03

Released on July 4th, 2019.

* Upgraded `oxauth-client` to 3.1.6.sp2.

## Version 3.1.6_02

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.6_01

Released on April 29th, 2019.

* Upgraded to Gluu Server 3.1.6.

## Version 3.1.5_04

Released on May 10th, 2019.

* Alpine upgraded to v3.9. Ref: https://github.com/GluuFederation/gluu-docker/issues/71.

## Version 3.1.5_03

Released on April 23rd, 2019.

* Fixed usage on KeyGenerator CLI call. Issue: https://github.com/GluuFederation/docker-key-rotation/issues/4.

## Version 3.1.5_02

Released on April 9th, 2019.

* Added license info on container startup.

## Version 3.1.5_01

Released on March 23rd, 2019.

* Upgraded to Gluu Server 3.1.5.

## Version 3.1.4_02

Released on March 15th, 2019.

* Fixed issue where `oxauth_key_rotated_at` is saved even when `oxauth_jks_base64` config couldn't be saved.

## Version 3.1.4_01

Released on November 12th, 2018.

* Upgraded to Gluu Server 3.1.4.

## Version 3.1.3_05

Released on September 18th, 2018.

* Changed base image to use Alpine 3.8.1.

## Version 3.1.3_04

Released on September 12th, 2018.

* Added feature to connect to secure Consul (HTTPS).

## Version 3.1.3_03

Released on August 31st, 2018.

* Added Tini to handle signal forwarding and reaping zombie processes.

## Version 3.1.3_02

Released on July 20th, 2018.

* Added wrapper to manage config via Consul KV or Kubernetes configmap.

## Version 3.1.3_01

Released on June 12th, 2018.

* Upgraded to Gluu Server 3.1.3

## Version 3.1.2_01

Released on June 12th, 2018.

* Upgraded to Gluu Server 3.1.2

## Version 3.1.1_rev1.0.0-beta2

Released on October 11th, 2017.

* Use latest oxauth-client build.

## Version 3.1.1_rev1.0.0-beta1

Released on October 6th, 2017.

* Migrated to Gluu Server 3.1.1.

## Version 3.0.1_rev1.0.0-beta2

Released on August 16th, 2017.

* Fixed base64 string when rotating keys.

## Version 3.0.1_rev1.0.0-beta1

Released on July 7th, 2017.

* Added feature to rotate oxAuth's public and private keys.


scripts/settings.py
LOGGING_CONFIG = {
    "version": 1,
    "formatters": {
        "default": {
            "format": "%(levelname)s - %(name)s - %(asctime)s - %(message)s",
        },
    },
    "handlers": {
        "console": {
            "class": "logging.StreamHandler",
            "formatter": "default",
        },
    },
    "loggers": {
        "pygluu.containerlib": {
            "handlers": ["console"],
            "level": "INFO",
            "propagate": True,
        },
        "entrypoint": {
            "handlers": ["console"],
            "level": "INFO",
            "propagate": False,
        },
        "wait": {
            "handlers": ["console"],
            "level": "INFO",
            "propagate": False,
        },
    },
    # "root": {
    #     "level": "INFO",
    #     "handlers": ["console"],
    # },
}


scripts/wait_for_it.py
#! /usr/bin/python2
# -*- coding: utf-8-unix -*-

# Checks waits for the following to happen before moving on to the
# passed command:
#
# - consul is up and populated
# - ldap is up and populated
#
# author: torstein@escenic.com

import base64
import ldap3
import logging as log
import os
import pyDes
import sys
import time

from gluu_config import ConfigManager

GLUU_LDAP_URL = os.environ.get("GLUU_LDAP_URL", "ldap:1636")

SLEEP_DURATION = 5
LAST_CONFIG_KEY = "oxauth_openid_jwks_fn"

try:
    GLUU_WAIT_MAX_TIME = int(os.environ.get("GLUU_WAIT_MAX_TIME", 300))
except ValueError:
    GLUU_WAIT_MAX_TIME = 300


def wait_for_config(config_manager):
    for i in range(0, GLUU_WAIT_MAX_TIME, SLEEP_DURATION):
        try:
            if config_manager.get(LAST_CONFIG_KEY):
                log.info("Config backend is ready.")
                return
        except Exception as exc:
            log.warn(exc)
            log.warn(
                "Config backend is not ready, retrying in {} seconds.".format(
                    SLEEP_DURATION))
        time.sleep(SLEEP_DURATION)

    log.error("Config backend is not ready after {} seconds.".format(GLUU_WAIT_MAX_TIME))
    sys.exit(1)


def get_ldap_password(consul):
    encoded_password = config_manager.get("encoded_ox_ldap_pw")
    encoded_salt = config_manager.get("encoded_salt")
    cipher = pyDes.triple_des(
        b"{}".format(encoded_salt),
        pyDes.ECB,
        padmode=pyDes.PAD_PKCS5
    )
    encrypted_text = b"{}".format(base64.b64decode(encoded_password))
    return cipher.decrypt(encrypted_text)


def wait_for_ldap(config_manager):
    ldap_bind_dn = config_manager.get("ldap_binddn")
    ldap_password = get_ldap_password(config_manager)

    ldap_host = GLUU_LDAP_URL.split(":")[0]
    ldap_port = int(GLUU_LDAP_URL.split(":")[1])

    ldap_server = ldap3.Server(
        ldap_host,
        ldap_port,
        use_ssl=True
    )
    log.warn(
        "LDAP trying ldaps://" + str(GLUU_LDAP_URL) +
        " ldap_bind_dn=" + ldap_bind_dn
    )

    # check the entries few times, to ensure OpenDJ is running after importing
    # initial data; this may not required for OpenLDAP
    successive_entries_check = 0

    for i in range(0, GLUU_WAIT_MAX_TIME, SLEEP_DURATION):
        try:
            with ldap3.Connection(
                    ldap_server,
                    ldap_bind_dn,
                    ldap_password) as ldap_connection:

                ldap_connection.search(
                    search_base="o=gluu",
                    search_filter="(oxScopeType=openid)",
                    search_scope=ldap3.SUBTREE,
                    attributes=['*']
                )

                if successive_entries_check >= 3:
                    log.info("LDAP is up and populated :-)")
                    return 0

                if ldap_connection.entries:
                    successive_entries_check += 1

        except Exception as exc:
            log.warn(
                "LDAP not yet initialised: {}; retrying in {} seconds".format(
                    exc, SLEEP_DURATION,
                )
            )
        time.sleep(SLEEP_DURATION)

    log.error("LDAP not ready, after " + str(GLUU_WAIT_MAX_TIME) + " seconds.")
    sys.exit(1)


def execute_passed_command(command_list):
    log.info(
        "Now executing the arguments passed to " +
        sys.argv[0] +
        ": " +
        " ".join(command_list)
    )
    os.system(" ".join(command_list))


def configure_logger():
    # When debugging wait-for-it, set level=log.DEBUG or pass
    # --log=DEBUG on the command line.
    log.basicConfig(
        level=log.INFO,
        format='%(asctime)s [%(levelname)s] [%(filename)s] - %(message)s'
    )


if __name__ == "__main__":
    configure_logger()

    log.info(
        "Hi world, waiting for config backend & LDAP to be ready before " +
        "running " + " ".join(sys.argv[1:])
    )
    config_manager = ConfigManager()
    wait_for_config(config_manager)
    wait_for_ldap(config_manager)
    execute_passed_command(sys.argv[1:])


scripts/entrypoint.py
import base64
import json
import logging
import os
import random
import shlex
import string
import subprocess
import sys
import time

import pyDes
from ldap3 import Connection
from ldap3 import Server
from ldap3 import BASE
from ldap3 import MODIFY_REPLACE
from ldap3.core.exceptions import LDAPSocketOpenError

from gluulib import get_manager

GLUU_LDAP_URL = os.environ.get("GLUU_LDAP_URL", "localhost:1636")

# Interval between rotation (in hours)
GLUU_KEY_ROTATION_INTERVAL = os.environ.get("GLUU_KEY_ROTATION_INTERVAL", 48)

# check interval (by default per 1 hour)
GLUU_KEY_ROTATION_CHECK = os.environ.get("GLUU_KEY_ROTATION_CHECK", 60 * 60)

# Default charset
_DEFAULT_CHARS = "".join([string.ascii_uppercase,
                          string.digits,
                          string.lowercase])

SIG_KEYS = "RS256 RS384 RS512 ES256 ES384 ES512"
ENC_KEYS = "RSA_OAEP RSA1_5"


logger = logging.getLogger("key_rotation")
logger.setLevel(logging.INFO)
ch = logging.StreamHandler()
fmt = logging.Formatter('%(levelname)s - %(asctime)s - %(message)s')
ch.setFormatter(fmt)
logger.addHandler(ch)

manager = get_manager()


def exec_cmd(cmd):
    args = shlex.split(cmd)
    popen = subprocess.Popen(args,
                             stdin=subprocess.PIPE,
                             stdout=subprocess.PIPE,
                             stderr=subprocess.PIPE)
    stdout, stderr = popen.communicate()
    retcode = popen.returncode
    return stdout, stderr, retcode


def generate_openid_keys(passwd, jks_path, dn, exp=365):
    if os.path.isfile(jks_path):
        os.unlink(jks_path)

    cmd = " ".join([
        "java",
        "-jar", "/opt/key-rotation/javalibs/keygen.jar",
        "-enc_keys", ENC_KEYS,
        "-sig_keys", SIG_KEYS,
        "-dnname", "{!r}".format(dn),
        "-expiration_hours", "{}".format(exp),
        "-keystore", jks_path,
        "-keypasswd", passwd,
    ])

    out, err, retcode = exec_cmd(cmd)
    return out, err, retcode


def should_rotate_keys():
    last_rotation = manager.config.get("oxauth_key_rotated_at")

    # keys are not rotated yet
    if not last_rotation:
        return True

    # when keys are supposed to be rotated
    next_rotation = int(last_rotation) + (60 * 60 * int(GLUU_KEY_ROTATION_INTERVAL))

    # current timestamp
    now = int(time.time())

    # check if current timestamp surpassed expected rotation timestamp
    return now > next_rotation


def decrypt_text(encrypted_text, key):
    cipher = pyDes.triple_des(b"{}".format(key), pyDes.ECB,
                              padmode=pyDes.PAD_PKCS5)
    encrypted_text = b"{}".format(base64.b64decode(encrypted_text))
    return cipher.decrypt(encrypted_text)


def modify_oxauth_config(ldap_conn, entry_dn, ox_rev, conf_dynamic, conf_webkeys):
    logger.info("modifying oxAuth configuration")

    serialized_keys_conf = json.dumps(conf_webkeys)
    serialized_dyn_conf = json.dumps(conf_dynamic)
    ox_rev = str(ox_rev + 1)

    ldap_conn.modify(entry_dn, {
        'oxRevision': [(MODIFY_REPLACE, [ox_rev])],
        'oxAuthConfWebKeys': [(MODIFY_REPLACE, [serialized_keys_conf])],
        'oxAuthConfDynamic': [(MODIFY_REPLACE, [serialized_dyn_conf])],
    })

    result = ldap_conn.result["description"]
    return result == "success"


def encode_jks(jks="/etc/certs/oxauth-keys.jks"):
    encoded_jks = ""
    with open(jks, "rb") as fd:
        encoded_jks = encrypt_text(fd.read(), manager.secret.get("encoded_salt"))
    return encoded_jks


def rotate_keys(user, passwd, inum, jks_pass, jks_fn, jks_dn):
    try:
        logger.info("connecting to server {}".format(GLUU_LDAP_URL))

        ldap_server = Server(GLUU_LDAP_URL, port=1636, use_ssl=True)

        with Connection(ldap_server, user, passwd) as conn:
            # get oxAuth config from LDAP
            ox_config = get_oxauth_config(conn, inum)

            if not ox_config:
                # search failed due to missing entry
                logger.warn("unable to find oxAuth config")
                return

            ox_rev = int(ox_config["oxRevision"][0])

            conf_dynamic = json.loads(ox_config["oxAuthConfDynamic"][0])

            if conf_dynamic["keyRegenerationEnabled"]:
                logger.warn("keyRegenerationEnabled config was set to true; "
                            "skipping proccess to avoid conflict with "
                            "builtin key rotation feature in oxAuth")
                return

            conf_dynamic.update({
                "keyRegenerationEnabled": False,  # always set to False
                "keyRegenerationInterval": int(GLUU_KEY_ROTATION_INTERVAL),
                "webKeysStorage": "keystore",
                "keyStoreSecret": jks_pass,
            })

            try:
                conf_webkeys = json.loads(ox_config["oxAuthConfWebKeys"][0])
            except IndexError:
                conf_webkeys = {"keys": []}

            exp_hours = int(GLUU_KEY_ROTATION_INTERVAL) + (conf_dynamic["idTokenLifetime"] / 3600)

            out, err, retcode = generate_openid_keys(
                jks_pass, jks_fn, jks_dn, exp=exp_hours)

            if retcode != 0:
                logger.error("unable to generate keys; reason={}".format(err))
                return

            try:
                new_keys = json.loads(out)
                merged_webkeys = merge_keys(new_keys, conf_webkeys)
                ox_modified = modify_oxauth_config(
                    conn, ox_config.entry_dn, ox_rev, conf_dynamic, merged_webkeys)

                if all([ox_modified,
                        manager.secret.set("oxauth_jks_base64", encode_jks())]):
                    manager.config.set("oxauth_key_rotated_at", int(time.time()))
                    logger.info("keys have been rotated")
            except (TypeError, ValueError) as exc:
                logger.warn("unable to get public keys; reason={}".format(exc))
    # cant connect to LDAP
    except LDAPSocketOpenError as exc:
        logger.warn("Unable to connect to LDAP at {}; reason={}".format(
            GLUU_LDAP_URL, exc))


def main():
    validate_rotation_check()
    validate_rotation_interval()

    inum = manager.config.get("inumAppliance")
    user = manager.config.get("ldap_binddn")
    passwd = decrypt_text(manager.secret.get("encoded_ox_ldap_pw"),
                          manager.secret.get("encoded_salt"))
    # jks_pass = get_random_chars()
    jks_pass = manager.secret.get("oxauth_openid_jks_pass")
    jks_fn = manager.config.get("oxauth_openid_jks_fn")
    jks_dn = r"{}".format(manager.config.get("default_openid_jks_dn_name"))

    try:
        while True:
            logger.info("checking whether key should be rotated")

            try:
                if should_rotate_keys():
                    rotate_keys(user, passwd, inum, jks_pass, jks_fn, jks_dn)
                else:
                    logger.info("no need to rotate keys at the moment")
            except Exception as exc:
                logger.warn("unable to connect to config backend; "
                            "reason={}".format(exc))
            time.sleep(int(GLUU_KEY_ROTATION_CHECK))
    except KeyboardInterrupt:
        logger.warn("canceled by user; exiting ...")


def encrypt_text(text, key):
    cipher = pyDes.triple_des(b"{}".format(key), pyDes.ECB,
                              padmode=pyDes.PAD_PKCS5)
    encrypted_text = cipher.encrypt(b"{}".format(text))
    return base64.b64encode(encrypted_text)


def get_random_chars(size=12, chars=_DEFAULT_CHARS):
    """Generates random characters.
    """
    return ''.join(random.choice(chars) for _ in range(size))


def get_oxauth_config(conn, inum):
    # base DN for oxAuth config
    oxauth_base = ",".join([
        "ou=oxauth",
        "ou=configuration",
        "inum={}".format(inum),
        "ou=appliances",
        "o=gluu",
    ])

    conn.search(
        search_base=oxauth_base,
        search_filter="(objectClass=*)",
        search_scope=BASE,
        attributes=[
            "oxRevision",
            "oxAuthConfWebKeys",
            "oxAuthConfDynamic",
        ]
    )

    if not conn.entries:
        return
    return conn.entries[0]


def merge_keys(new_keys, old_keys):
    """Merges new and old keys while omitting expired key.
    """
    now = int(time.time() * 1000)
    for key in old_keys["keys"]:
        if key.get("exp") > now:
            new_keys["keys"].append(key)
    return new_keys


def validate_rotation_check():
    err = "GLUU_KEY_ROTATION_CHECK must use a valid integer greater than 0"
    try:
        if int(GLUU_KEY_ROTATION_CHECK) < 1:
            logger.error(err)
            sys.exit(1)
    except ValueError:
        logger.error(err)
        sys.exit(1)


def validate_rotation_interval():
    err = "GLUU_KEY_ROTATION_INTERVAL must use a valid integer greater than 0"
    try:
        if int(GLUU_KEY_ROTATION_INTERVAL) < 1:
            logger.error(err)
            sys.exit(1)
    except ValueError:
        logger.error(err)
        sys.exit(1)


if __name__ == "__main__":
    main()
